frontend/src/components/samples/details/SampleDetailsContent.js

```diff
@@ -139,47 +112,56 @@ const SampleDetailsContent = ({
 
 
   // Navigate to a sample when the sample's node is clicked in the lineage graph.
-  const navigateToSample = (sample_id) => {
+  const navigateToSample = useCallback((sample_id) => {
     if (sample_id) {
       history(`/samples/${sample_id}#lineage`)
     }
-  }
+  }, [history])
 
   // Navigate to a process measurement when the user clicks an edge
   // in the lineage graph.
-  const navigateToProcess = (process_id) => {
+  const navigateToProcess = useCallback((process_id) => {
     if (process_id) {
       history(`/process-measurements/${process_id}`)
     }
-  }
+  }, [history])
 
   // TODO: This spams API requests
-  if (!samplesByID[id])
-    getSample(id);
+  useEffect(() => {
+    if (!samplesByID[id])
+      dispatch(getSample(id));
+  }, [dispatch, id, samplesByID])
 
-  if (isLoaded && !sample.versions && !sample.isFetching)
-    listVersions(sample.id);
+  useEffect(() => {
+    if (isLoaded && !sample.versions && !sample.isFetching)
+      dispatch(listVersions(sample.id));
+  }, [dispatch, isLoaded, sample.id, sample.isFetching, sample.versions])
 
-  if (isLoaded && !isProcessesEmpty) {
-    sample.process_measurements.forEach((id, i) => {
-      withProcessMeasurement(processMeasurementsByID, id, process => process.id);
-      processMeasurements.push(processMeasurementsByID[id]);
-    })
-  }
+  useEffect(() => {
+    if (isLoaded && !isProcessesEmpty) {
+      fetchProcessMeasurements(sample.process_measurements).then((processMeasurements) => {
+        setProcessMeasurements(processMeasurements)
+      })
+    }
+  }, [isLoaded, isProcessesEmpty, sample.process_measurements])
 
-  if (isLoaded && container?.experiment_run) {
-    experimentRunsIDs.push(container.experiment_run)
-  }
+  useEffect(() => {
+    if (isLoaded && container?.experiment_run) {
+      setExperimentRunIDs((experimentRunsIDs) => [...experimentRunsIDs, container.experiment_run])
+    }
+  }, [container?.experiment_run, experimentRunsIDs, isLoaded])
 
-  if (!librariesByID[id])
-    getLibrary(id)
+  useEffect(() => {
+    if (!librariesByID[id])
+      dispatch(getLibrary(id))
+  }, [dispatch, id, librariesByID])
 
   useEffect(() => {
     const biosampleId = sample?.biosample_id
     listSampleMetadata(token, { "biosample__id": biosampleId }).then(metadata => {
       setSampleMetadata(metadata)
     })
-  }, [sample])
+  }, [sample, token])
```

Comment:
do you really want to refresh the sample metadata each time  the token expires and is getting replaced ?